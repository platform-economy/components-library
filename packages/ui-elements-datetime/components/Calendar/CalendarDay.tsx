import * as React from 'react';
import { Moment } from 'moment';
import { noop } from 'lodash';
import { weekDaySymbolicNames } from './Calendar.common';
import classnames from 'classnames';

export type CalendarDayProps = {
  date: Moment;
  isCurrentMonth: boolean;
  isCurrentDay: boolean;
  isSelectedDay: boolean;
  isSelectionStart: boolean;
  isSelectionEnd: boolean;
  isSelectionSpan: boolean;
  isDisabled?: boolean;
  onClick?: (date: Date, ev: React.SyntheticEvent) => void,
  onMouseEnter?: (date: Date, ev: React.SyntheticEvent) => void,
  onMouseLeave?: (date: Date, ev: React.SyntheticEvent) => void,
};

export class CalendarDay extends React.Component<CalendarDayProps> {

  private handleClick = (ev: React.SyntheticEvent) => {
    const { date, onClick = noop, isDisabled } = this.props;
    if (!isDisabled) {
      onClick(date.toDate(), ev);
    }
  }

  private handleMouseEnter = (ev: React.SyntheticEvent) => {
    const { date, onMouseEnter = noop, isDisabled } = this.props;
    if (!isDisabled) {
      onMouseEnter(date.toDate(), ev);
    }
  }

  private handleMouseLeave = (ev: React.SyntheticEvent) => {
    const { date, onMouseLeave = noop, isDisabled } = this.props;
    if (!isDisabled) {
      onMouseLeave(date.toDate(), ev);
    }
  }

  render() {
    const {
      date,
      isCurrentDay,
      isSelectedDay,
      isDisabled,
      isCurrentMonth,
      isSelectionStart,
      isSelectionEnd,
      isSelectionSpan,
    } = this.props;
    const localeData = date.localeData();
    const firstDayOfWeek = localeData.firstDayOfWeek();
    const weekdayIndex = (date.weekday() + firstDayOfWeek) % weekDaySymbolicNames.length;
    const weekdaySymbolicName = weekDaySymbolicNames[weekdayIndex];

    return (
      <td
        className={classnames('CalendarDay', 'day', `day-${weekdaySymbolicName}`, {
          current: isCurrentDay,
          selected: isSelectedDay || isSelectionStart || isSelectionEnd || isSelectionSpan,
          disabled: isDisabled,
          'current-month': isCurrentMonth,
          'other-month': !isCurrentMonth,
          'selection-single': isSelectedDay,
          'selection-start': isSelectionStart,
          'selection-end': isSelectionEnd,
          'selection-span': isSelectionSpan,
        })}
        onClick={this.handleClick}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {date.date()}
      </td>
    );
  }
}
