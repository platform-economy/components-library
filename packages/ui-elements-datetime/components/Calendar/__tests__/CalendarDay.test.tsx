import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import 'jest-styled-components';
import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { CalendarDay, CalendarDayProps } from '../CalendarDay';
import moment from 'moment';

const currentDate = new Date(2019, 5, 12, 0, 0, 0);

enzyme.configure({ adapter: new Adapter() });

function createProps(overrides: Partial<CalendarDayProps> = {}): CalendarDayProps {
  return {
    date: moment(currentDate),
    isCurrentMonth: true,
    isCurrentDay: false,
    isSelectedDay: false,
    isSelectionStart: false,
    isSelectionEnd: false,
    isSelectionSpan: false,
    isDisabled: false,
    ...overrides,
  };
}

describe('ui-elements-datetime', () => {
  describe('Components', () => {
    describe('CalendarDay', () => {

      const SNAPSHOT_TEST_CASES: {name: string, props: Partial<CalendarDayProps>}[] = [
        {
          name: 'default',
          props: {},
        },
        {
          name: 'as current day',
          props: { isCurrentDay: true },
        },
        {
          name: 'as single selection',
          props: { isSelectedDay: true },
        },
        {
          name: 'as selection range start',
          props: { isSelectionStart: true },
        },
        {
          name: 'as selection range end',
          props: { isSelectionEnd: true },
        },
        {
          name: 'as selection range span',
          props: { isSelectionSpan: true },
        },
        {
          name: 'as disabled',
          props: { isDisabled: true },
        },
      ];
      SNAPSHOT_TEST_CASES.forEach((testCase) => {
        it(`should render: ${testCase.name}`, () => {
          const props = createProps(testCase.props);
          const tree = TestRenderer
            .create(<CalendarDay {...props}/>)
            .toJSON();
          expect(tree).toMatchSnapshot();
        });
      });

      it('should handle click', () => {
        const props = createProps({
          onClick: jest.fn(),
        });
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('click', new Event('click'));
        const expectedDate = props.date.toDate();
        expect(props.onClick).toBeCalledTimes(1);
        expect(props.onClick).toBeCalledWith(expectedDate, expect.anything());
      });

      it('should handle click (no event attached)', () => {
        const props = createProps();
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('click', new Event('click'));
      });

      it('should handle mouse enter', () => {
        const props = createProps({
          onMouseEnter: jest.fn(),
        });
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('mouseenter', new Event('mouseenter'));
        const expectedDate = props.date.toDate();
        expect(props.onMouseEnter).toBeCalledTimes(1);
        expect(props.onMouseEnter).toBeCalledWith(expectedDate, expect.anything());
      });

      it('should handle mouse enter (no event attached)', () => {
        const props = createProps();
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('mouseenter', new Event('mouseenter'));
      });

      it('should handle mouse leave', () => {
        const props = createProps({
          onMouseLeave: jest.fn(),
        });
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('mouseleave', new Event('mouseleave'));
        const expectedDate = props.date.toDate();
        expect(props.onMouseLeave).toBeCalledTimes(1);
        expect(props.onMouseLeave).toBeCalledWith(expectedDate, expect.anything());
      });

      it('should handle mouse leave (no event attached)', () => {
        const props = createProps();
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('mouseleave', new Event('mouseleave'));
      });

      it('should not call click on disabled day', () => {
        const props = createProps({
          isDisabled: true,
          onClick: jest.fn(),
        });
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('click', new Event('click'));
        expect(props.onClick).toBeCalledTimes(0);
      });

      it('should not call mouse enter on disabled day', () => {
        const props = createProps({
          isDisabled: true,
          onMouseEnter: jest.fn(),
        });
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('mouseenter', new Event('mouseenter'));
        expect(props.onMouseEnter).toBeCalledTimes(0);
      });

      it('should not call mouse leave on disabled day', () => {
        const props = createProps({
          isDisabled: true,
          onMouseLeave: jest.fn(),
        });
        enzyme
          .shallow(<CalendarDay {...props}/>)
          .simulate('mouseleave', new Event('mouseleave'));
        expect(props.onMouseLeave).toBeCalledTimes(0);
      });
    });
  });
});
