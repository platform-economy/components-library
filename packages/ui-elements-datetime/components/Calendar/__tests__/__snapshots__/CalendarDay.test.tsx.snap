// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-elements-datetime Components CalendarDay should render: as current day 1`] = `
<td
  className="CalendarDay day day-wed current current-month"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;

exports[`ui-elements-datetime Components CalendarDay should render: as disabled 1`] = `
<td
  className="CalendarDay day day-wed disabled current-month"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;

exports[`ui-elements-datetime Components CalendarDay should render: as selection range end 1`] = `
<td
  className="CalendarDay day day-wed selected current-month selection-end"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;

exports[`ui-elements-datetime Components CalendarDay should render: as selection range span 1`] = `
<td
  className="CalendarDay day day-wed selected current-month selection-span"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;

exports[`ui-elements-datetime Components CalendarDay should render: as selection range start 1`] = `
<td
  className="CalendarDay day day-wed selected current-month selection-start"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;

exports[`ui-elements-datetime Components CalendarDay should render: as single selection 1`] = `
<td
  className="CalendarDay day day-wed selected current-month selection-single"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;

exports[`ui-elements-datetime Components CalendarDay should render: default 1`] = `
<td
  className="CalendarDay day day-wed current-month"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  12
</td>
`;
